params/config.go

```diff
@@ -171,6 +188,142 @@ type UpgradeConfig struct {
 	PrecompileUpgrades []PrecompileUpgrade `json:"precompileUpgrades,omitempty"`
 }
 
+func (c *UpgradeConfig) Hash() (common.Hash, error) {
+	bytes, err := c.MarshalBinary()
+	if err != nil {
+		return common.Hash{}, err
+	}
+	return crypto.Keccak256Hash(bytes), nil
+}
+
+func (c *UpgradeConfig) MarshalBinary() ([]byte, error) {
+	p := wrappers.Packer{
+		Bytes:   []byte{},
+		MaxSize: MaxMessageSize,
+	}
+	p.PackFixedBytes(magicHeader)
+	if p.Err != nil {
+		return nil, p.Err
+	}
+
+	p.PackBool(c.OptionalNetworkUpgrades == nil)
+	if p.Err != nil {
+		return nil, p.Err
+	}
+	if c.OptionalNetworkUpgrades != nil {
+		bytes, err := c.OptionalNetworkUpgrades.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		p.PackBytes(bytes)
+		if p.Err != nil {
+			return nil, p.Err
+		}
+	}
+
+	for _, precompileConfig := range c.PrecompileUpgrades {
+		bytes, err := precompileConfig.Config.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		p.PackFixedBytes(precompileHeader)
+		if p.Err != nil {
+			return nil, p.Err
+		}
+		p.PackStr(precompileConfig.Key())
+		if p.Err != nil {
+			return nil, p.Err
+		}
+		p.PackBytes(bytes)
+		if p.Err != nil {
+			return nil, p.Err
+		}
+	}
+
+	for _, config := range c.StateUpgrades {
+		bytes, err := config.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		p.PackFixedBytes(stateUpdateHeader)
+		if p.Err != nil {
+			return nil, p.Err
+		}
+		p.PackBytes(bytes)
+		if p.Err != nil {
+			return nil, p.Err
+		}
+	}
+
+	p.PackFixedBytes(endHeader)
+
+	return p.Bytes, nil
+}
+
+func (c *UpgradeConfig) UnmarshalBinary(data []byte) error {
+	p := wrappers.Packer{
+		Bytes: data,
+	}
+	if !bytes.Equal(p.UnpackFixedBytes(sectionHeaderSizes), magicHeader) {
+		return ErrIncorrectHeader
+	}
+	if p.Err != nil {
+		return p.Err
+	}
+
+	isNil := p.UnpackBool()
+	if !isNil {
+		c.OptionalNetworkUpgrades = &OptionalNetworkUpgrades{}
+		bytes := p.UnpackBytes()
+		if p.Err != nil {
+			return p.Err
+		}
+		if err := c.OptionalNetworkUpgrades.UnmarshalBinary(bytes); err != nil {
+			return nil
+		}
+	}
+
+	for {
```

Comment:
Let's switch to packing the number of elements of each type, so that we can proceed through field by field and unpack the number of elements, all of the elements of that type, and then proceed to the next field.

Reply:
Fixed in 8935259f